lib/streaming-pipeline.js

```diff
@@ -0,0 +1,193 @@
+        this._hookTimeoutInMs = options['hookTimeoutInMs'] || 10000;
+        this._userFunction = userFunction;
+        this._destinationStream = destinationStream;
+        this._startReceived = false;
+        this._functionInputs = [];
+        this._functionOutputs = [];
+        this._finishedOutputs = 0;
+        this.on('finish', () => {
+            logger('Ending input streams');
+            this._endInputs();
+        });
+        this.on('error', () => {
+            logger('Error occurred - ending pipeline now');
+            this.end();
+        });
+        this._invokeHooks();
+        logger('Streaming pipeline initialized');
+    }
+
+    _invokeHooks() {
+        const hookResolve = (hookName) => {
+            return (timerId) => {
+                clearTimeout(timerId);
+                logger(`${hookName} hook called`)
+            }
+        };
+        const emitErrorOnReject = (err) => {
+            this.emit('error', err)
+        };
+        if (typeof this._userFunction['$init'] === 'function') {
+            logger(`Calling $init hook with a max timeout of ${this._hookTimeoutInMs}ms `);
+            guardWithTimeout(this._userFunction['$init'], this._hookTimeoutInMs)
+                .then(hookResolve('$init'))
+                .catch(emitErrorOnReject);
+        }
+        if (typeof this._userFunction['$destroy'] === 'function') {
+            this._destinationStream.on('finish', () => {
+                logger(`Calling $destroy hook with a max timeout of ${this._hookTimeoutInMs}ms `);
+                guardWithTimeout(this._userFunction['$destroy'], this._hookTimeoutInMs)
+                    .then(hookResolve('$init'))
+                    .catch(emitErrorOnReject);
+            });
+        }
+    }
+
+    _endInputs() {
+        this._functionInputs.forEach((fi) => fi.emit('end'));
+    }
+
+    _write(inputSignal, _, callback) {
+        logger('Input signal received');
+        if (!inputSignal || !inputSignal['hasStart'] || !inputSignal['hasData']) {
+            callback(new RiffError(
+                'error-streaming-input-invalid',
+                `invalid input type ${Object.prototype.toString.call(inputSignal)}`));
+            return;
+        }
+        if (!inputSignal.hasStart() && !inputSignal.hasData()) {
+            callback(new RiffError(
+                'error-streaming-input-invalid',
+                'input is neither a start nor a data signal'));
+            return;
+        }
+
+        if (inputSignal.hasStart()) {
+            const outputContentTypes = inputSignal.getStart().getExpectedcontenttypesList();
+            if (this._startReceived) {
+                callback(new RiffError(
+                    'error-streaming-too-many-starts',
+                    `start signal has already been received. Rejecting new start signal with content types [${outputContentTypes.join()}]`));
+                return;
+            }
+            const outputCount = outputContentTypes.length;
```

Comment:
do we need this?

Concatenating the input and output stream arrays as function arguments is fraught. Even if the totals match who's to say the the function has the correct ratio of inputs and outputs?

Reply:
It's not perfect, but it's better than nothing, isn't it?